### ISSUE TITLE
Stack does not file-watch dependencies anymore

### ISSUE THREAD

**Issue by NorfairKing** [commercialhaskell/stack#5428]
### Steps to reproduce

Make a project with two packages: `a` and `b` where `b` is a dependency of `a`.
Start `stack install --file-watch a`.
Make a change in `b`.

This is also broken for intra-package dependencies.
For example when the executable depends on the library.

### Expected

A rebuild

### Actual

No rebuild

### Stack version

```
stack --version
2.5.1 x86_64 hpack-0.34.2
```

### Method of installation

* nixpkgs

**Comment by NorfairKing** [comment]
Workaround: `stack install a b --file-watch`

**Comment by NorfairKing** [comment]
It looks like this was intentional; https://github.com/commercialhaskell/stack/issues/5310
I'm not entirely sure whether this change was entirely intentional because watching dependencies seems like a good default while watching unrelated packages does not seem like a good default.

In any case, another work around based on that issue: `stack install a --watch-all`.

**Comment by qrilka** [comment]
@NorfairKing I suppose that you want to have dependencies taken into consideration for file watching. In principle that wasn't available before #5317 but every local file was watched so it looked like it was. That option is still available with `--watch-all`. Adding files from local dependencies into watch list requires a new PR.

**Comment by dten** [comment]
@NorfairKing My intention with #5317 was just not not watch watch files from unrelated packages / components. I just got round to upgrading and it seems it didn't quite work out how I imagined it would😅

**Comment by qrilka** [comment]
@dten please raise your concerns if you think that #5317 results in some unexpected behavior

**Comment by dten** [comment]
Hey yea so my actual life project has a structure roughly like this

````
stack.yaml
  - project webhost
    - webhost-lib
    - webhost-executable
      - dep: webhost
      - dep: common_stuff
  - project tester
    - tester-lib
    - tester-tests
      - dep: tester
      - dep: common_stuff
  - project common_stuff
    - common_stuff-lib
````

the webhost runs and is auto reloaded by yesod-bin
then (in another stack-work) the tester project can run tests against the running host

before #5317 watch building the webhost would watch _all_ libs, which meant that changing the common lib would cause a rebuild (desirable) but also changing the test lib would cause a rebuild (undesirable).

it seems now that _only_ the webhost lib is watched, so changing the common stuff does not result in rebuilding (undesirable)

my intention was that since common_stuff-lib is part of the local dependencies of the build target(s) then it should also be watched, but since tester-lib is not then it should not.

**Comment by dten** [comment]
please note I also mean if common_stuff-lib had local deps they would also be watched

**Comment by qrilka** [comment]
Yeah @dten it looks like you want package dependencies taken into consideration - that makes sense. I will try to find some time to look into it. It doesn't look to be a hard one.

**Comment by dten** [comment]
awesome thanks

**Comment by dten** [comment]
I take it this is on the back burner?

**Comment by qrilka** [comment]
Unfortunately yes - to implement this we need to construct extra dependency graph and I didn't find time to work on this

**Comment by dten** [comment]
is this something I might be able to do with some guidance ?

**Comment by qrilka** [comment]
That would be great but unfortunately will be mostly away on my vacation for the next 2 weeks so if you want we could restart this discussion somewhere around the 1st of August. Thanks
